### ISSUE TITLE
Rework quiver's color-setting API

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#13423]
quiver's docs currently read
```
    Call signatures::
    
      quiver(U, V, **kw)
      quiver(U, V, C, **kw)
      quiver(X, Y, U, V, **kw)
      quiver(X, Y, U, V, C, **kw)
    
    *U* and *V* are the arrow data, *X* and *Y* set the location of the
    arrows, and *C* sets the color of the arrows. These arguments may be 1-D or
    2-D arrays or sequences.

    ...

    Parameters
    ----------
    X : 1D or 2D array, sequence, optional
        The x coordinates of the arrow locations
    Y : 1D or 2D array, sequence, optional
        The y coordinates of the arrow locations
    U : 1D or 2D array or masked array, sequence
        The x components of the arrow vectors
    V : 1D or 2D array or masked array, sequence
        The y components of the arrow vectors
    C : 1D or 2D array, sequence, optional
        The arrow colors
```

OK, let's try to draw 4 arrows, each with u=v=1, in red (1, 0, 0), green (0, 1, 0), blue (0, 0, 1), and gray (.5, .5, .5)
```
quiver([1, 1, 1, 1], [1, 1, 1, 1], ["r", "g", "b", "gray"])
...
TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```
Oops.  Let's try to pass in RGB color tuples instead? (in the same format as `scatter` would take them)
```
quiver([1, 1, 1, 1], [1, 1, 1, 1], [[1, 0, 0], [0, 1, 0], [0, 0, 1], [.5, .5, .5]])
```
![bad](https://user-images.githubusercontent.com/1322974/52657727-9307b100-2ef9-11e9-9f03-aec9ed1b504a.png)
??
To cut a long story short: this took the first 4 values in the flattened color array (1, 0, 0, 0) and colormapped them using the default colormap.

Actually, the correct solution to set the colors is to read all the way to the end of the docstring and notice
```
    color : [ color | color sequence ], optional
        This is a synonym for the
        :class:`~matplotlib.collections.PolyCollection` facecolor kwarg.
        If *C* has been set, *color* has no effect.
```
Indeed
```
quiver([1, 1, 1, 1], [1, 1, 1, 1], color=["r", "g", "b", "gray"])
```
or
```
quiver([1, 1, 1, 1], [1, 1, 1, 1], color=[[1, 0, 0], [0, 1, 0], [0, 0, 1], [.5, .5, .5]])
```
both work:
![good](https://user-images.githubusercontent.com/1322974/52657714-8edb9380-2ef9-11e9-8cd5-6f81d1f39c70.png)

I think this API is clearly... not optimal.

I propose to instead make C check its shape; if its shape is `(*X.shape, 3)` or `(*X.shape, 4)`, then treat it as RGB; if its shape is `X.shape`, treat it as colormapped values; otherwise, error out (and for consistency, check that the shapes of X, Y, U, V all match (instead of flattening them all)).

I'm not sure the behavior change needs a deprecation period (it would only affect people purposefully passing three or four times to many values as the C array as the rest would currently be ignored); obviously there would be a deprecation during which non-matching X, Y, U, V would warn instead of error out.

The color arg could also be deprecated, as it is now a complete synonym for C.  Alternatively, we could instead deprecate C in favor of color, which would have the benefit of making the signatures a bit nicer (now it's just `quiver(U, V, **kwargs)` and `quiver(X, Y, U, V, **kwargs)`) but is perhaps mildly more disruptive.

Thoughts?

This is as of mpl 3.0.x.

**Comment by timhoffm** [comment]
Yep, that API is not optimal (and I did dare to update the documentation so far because of this).

So the argument `C` is currently not working at all? From a pure practical point of view, this would make it easier to deprecate since there should not be any working user code.

I'm a bit lost with respect to API consistency:
- [pcolormesh](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.pcolormesh.html#matplotlib.axes.Axes.pcolormesh), [pcolorfast](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.pcolorfast.html#matplotlib.axes.Axes.pcolorfast) and [pcolor](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.pcolor.html#matplotlib.axes.Axes.pcolor) all use a positional `C` argument.
- [tripcolor](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.tripcolor.html#matplotlib.axes.Axes.tripcolor) uses a hybid of `C` and `facecolor`.
- [scatter](https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.tripcolor.html#matplotlib.axes.Axes.tripcolor) has a lowercase `c` and some color kwargs.

Not really thinking this completely through, but `quiver` is semantically most similar to `scatter`.
Generally, I don't think that we suport standard Artist kwags to be arrays of args, though there might be exceptions. That might be a reason to keep `C` or `c` arguments.

**Comment by anntzer** [comment]
`C` is working for setting scalar values that will be normalized and colormapped.  So I guess the question is whether we want to overload the same argument to mean "either scalar values that will be colormapped, or RGB(A) values".  This is consistent with scatter, and there is no risk of confusion as the shapes clearly indicate which is the case (... unless we flatten everything and ignore mismatching shapes, as we do right now :/).
So I guess we can rename `C` to `c` (right now it's positional only so its name is not part of the API; `c` is consistent with `scatter`) and keep `color` as a synonym? (It's a standard synonym anyways, e.g. for `plot()`; if we decide one day to deprecate shorthands (not suggesting this here) having it in quiver() won't make the deprecation easier or harder.)
I don't think the (tri)pcolor(foo) family is directly comparable as for them the color is the most important thing, whereas for quiver and scatter it's just one possible way to decorate (in a nontechnical sense) the artist.

**Comment by efiring** [comment]
It sounds like you are on the right track, but regarding the overloading, wouldn't you want to include a sequence of mpl colorspecs rather than just the RGB(A) case?

**Comment by anntzer** [comment]
Yup, basically it should be whatever to_rgba_array is willing to handle (after flattening the x/y).

**Comment by dopplershift** [comment]
Does #14424 fix this?

**Comment by anntzer** [comment]
Now the examples above raise errors, which is an improvement over the earlier situation; the remaining question is whether they should indeed raise errors or be valid forms to set the colors with the semantics suggested above.
